File: 0320. Generalized Abbreviation.java
// Medium
// Backtracking
// Time:O(n*2^n),Space:O(n)
// https://leetcode.cn/problems/generalized-abbreviation/

import java.util.*;

class Solution {
    public List<String> generateAbbreviations(String word) {
        List<String> res = new ArrayList<>();
        backtracking(res, new StringBuilder(), word, 0, 0); // index, count
        return res;
    }

    private void backtracking(List<String> res, StringBuilder abbr, String word, int index, int count) {
        int n = abbr.length();

        if (index == word.length()) {
            if (count > 0) abbr.append(count);
            res.add(abbr.toString());
        } else {
            backtracking(res, abbr, word, index + 1, count + 1);

            // for each char, can choose skip or not
            // if no skip cur char, deal with the count first
            if (count > 0) abbr.append(count);
            abbr.append(word.charAt(index)); // than add cur char
            backtracking(res, abbr, word, index + 1, 0);
        }
        abbr.setLength(n);
    }
}

File: 0001-0500/0320. Generalized Abbreviation.py
# Time:O(n * 2^n), Space:O(n)

from typing import List

class Solution:
    def generateAbbreviations(self, word: str) -> List[str]:
        res = []

        def backtracking(index, path, count):
            if index == len(word):
                res.append(path + (str(count) if count > 0 else ""))
                return
            
            # cnt + 1
            backtracking(index + 1, path, count + 1)

            # keep current
            backtracking(
                index + 1, path + (str(count) if count > 0 else "") + word[index], 0
            )

        backtracking(0, "", 0)
        return res
    